### ISSUE TITLE
OneDrive issues

### ISSUE THREAD

**Issue by gregdavisd** [pwsafe/pwsafe#1002]
It causes constant error with Windows OneDrive sync. "There is a sync issue in OneDrive"  *****.plk "This file is in use. This file is open by another program and can't be synced. Once the file is closed, we can sync it." I assume this is some kind of lock file. It freezes one drive. I've turned off backup database files. OneDrive has file versioning so no point doing that. But still causes issue with OneDrive. I don't see why need to create a file to indicate a file is locked. Maybe it's a Linux thing. But I use a real operating system.

**Comment by rafaelx** [comment]
The lock file is pwsafe specific and is used by the application to indicate that the database is already open by a pwsafe instance.
In fact, on Windows (if that's what you call a real operating system) the lock file cannot be synced from OneDrive as long as it is open from pwsafe.
In my case, OneDrive only signals that a synchronization task is still pending. However, other synchronization tasks will continue to work, so the database will also be correctly synchronized, as will newly created files by other applications. At least the plk file doesn't block any other OneDrive tasks for me.

**Comment by gregdavisd** [comment]
I don't see how I'm going to get into a race condition against myself. It shouldn't be multi user synchronized given that it's a database of personal passwords. Using save only after a change means the file won't be open for the entire duration. The PLK file is going for me. Once I can figure out how to build it. VS Code doesn't work at all and VS Community 2022 says project is broken cannot import "...pwsafe\src\ui\Windows\UserVariables-15.props". Also another bug, open PWSafe from the notification area and it doesn't grab window focus. Start typing password and just getting a dinging noise as Windows doesn't know where to send the keystrokes. So have to move mouse and click on window after already opening up notification area and clicking the icon there.

**Comment by gregdavisd** [comment]
I figured it out. I ran "configure-16.vbs" but that is just a decoy. The entire file contents is:

{
	"folders": []
}

Which does nothing

Run configure-15.vbs instead. Then to use the 16 SLN file need to rename it to UserVariables-16.props

**Comment by gregdavisd** [comment]
I was able to open the same database file on both desktop and android. Both in read/write mode and save conflicting copies. Later I got a warning on OneDrive that there were two conflicting copies. So the .plk file doesn’t do much on OneDrive. Actually the .plk file doesn’t show up anywhere other than one device that created it. I waited 25 mins and nothing. This could be because the .plk file is locked and can’t be synced or OneDrive is just taking its sweet time getting around to uploading the .plk file. OneDrive doesn’t network share files, it creates a copy on each device and syncs it later. Making any attempt at real time synchronization futile. Only way to do that would be with middleware.
I’m going to instead use the OS to monitor and generate a file changed event. This is how NotePad++ does it. It won’t be perfect, just have to use the versioning feature to sort it out later.

**Comment by Chiramisu** [comment]
TL;DR: Move the plk to `%LocalAppData%\PasswordSafe` on Windows and call it a day. It shouldn't be syncing anyway. 😏

**Proof**: I'm on v3.62 on two PCs, with psafe3 file synced on OneDrive. I was able to overwrite the database on one machine with the contents from another by doing the following:

1. Close all DB files on both machines
2. Create new test DB on Machine A.
3. Create new credential.
4. Close and ensure that it syncs to Machine B.
5. Open again on Machine A.
6. Add a digit to password on test credential.
7. Open on Machine B while still open on Machine A.
  **Note**: Since the plk doesn't sync over OneDrive when the DB is open, it doesn't prevent the DB from being open on other machines.
8. Add another test credential on Machine B.
9. Save and Close on Machine B.
10. Wait a moment, then Save and Close on Machine A.
11. Wait for OneDrive to fully sync on both Machines.
12. Open again on Machine B and note that the changes from Machine A overwrote the next credential you created on Machine B, and that new credential was lost.
  **Note**: The lost credential can be recovered from an Intermediate backup.

Bottom line: if the purpose of the plk is to prevent the DB from being overwritten, it is not able to prevent this when the DB is updated simultaneously on two separate machines which are synced by third-party software. You may consider this an edge case, but I would argue that it's not all that uncommon. I have gotten in the habit of Closing my DBs when not in use because of this issue, but before I would frequently leave the DB open on my desktop and put it to sleep (no syncing), but then do some work on my laptop and end up needing a password, then updating expired ones, and not closing it on that machine either. If I were to close the DB on my desktop first, then close it on my laptop and re-open it after OneDrive synced, I would lose my updated credentials. Another cruel defect I noticed, which I'll probably file separately is that when you backup from an Intermediate backup, the others get blown away, so if you choose the wrong one, you're screwed. Except for the fact that OneDrive has backups of those as well.

**Comment by gregdavisd** [comment]
I've had sync issues with PasswordSafe. While roaming I changed passwords on my phone app. It updated later or didn't. Lost access to a password. Not sure the steps but the new password I generated was gone for good. One thing is. My entire password ecosystem does rely on only a few passwords. The passwords to my emails. Using my emails I can reset any password. So my email passwords act as a sort of master password. I don't see a way to have perfect synchronization when users are constantly switching devices and connectivity. Maybe a blockchain database of passwords? A Merkle tree? I've learnt to keep those few key passwords safe and that's it. Use email password reset in case something goes wrong.

**Comment by gregdavisd** [comment]
It happened again. So rare that is now far in to the future. I got "conflicted copy..." on OneDrive. What I did was open up on Android with OneDrive cloud sync, set to writable, made my changes and left it for days. Then I open up on Windows and I get conflicted copy. The conflicted copy was the old version before changing on Android. I think when OneDrive is in use there should be no ".plk" lock file or backups. Because OneDrive already does backups and doesn't allow locks anyway. I was able to recover by keeping the original and the conflicted copy as recommended by OneDrive. At the end of the day, in the end, after all, a filesystem doesn't offer ACID database compliance. There will always be conflicts when multiple devices are involved. So how to avoid this issue? As I've said, It's a Microsoft/OneDrive problem. Maybe just a notice when toggling the "writable" option that explains to be careful.

**Comment by gregdavisd** [comment]
The PasswordSafe Sync app on Android is set to every hour. It was 3 days. When it happens it makes me question reality itself. I'm at the bar trying to pay for the liquid gold just poured into my glass but I can't enter my card PIN. So I try to go to my Online banking but the password doesn't work. It's punishing. I know that if I sit at two devices at once and edit, obviously, I'm going to have issues. I didn't expect a 3 day window to cause this issue. I had to excuse myself while I sorted it out. It was recoverable, just pay cash. When I get home on Windows I find conflicted copies. I'm saying that all these attempts to synchronize between Android/Windows or multiple devices don't really work in the case of OneDrive. The original problem was the ".plk" file always getting stuck synchronizing with OneDrive. For a lock file to act as a flag to lock the file then it has to be visible to all devices, which it's not. The plk file never syncs until PasswordSafe is shutdown on which it's deleted anyway.

**Comment by gregdavisd** [comment]
I guess it's now a feature request. The cloud syncing will never work with an 'rsync' type protocol. Instead host an E2E encrypted service that costs $5.99 a month (or there about) for a server. Obviously one can deploy their own server with the open source, and if they have their own server. Otherwise the edge cases on these sync issues will be never ending. It's 2025 (current year argument) and if people are handing over cash for subscriptions for this that and the other then, why not?
